plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -152,53 +226,181 @@ public NuageVspDeviceVO addNuageVspDevice(AddNuageVspDeviceCmd cmd) {
+        String apiVersion = Objects.firstNonNull(command.getApiVersion(), getClientDefaultsAnswer.getCurrentApiVersion());
+        SupportedApiVersionCommand supportedApiVersionCmd = new SupportedApiVersionCommand(apiVersion);
+        SupportedApiVersionAnswer supportedApiVersionAnswer = (SupportedApiVersionAnswer) _agentMgr.easySend(nuageVspHost.getId(), supportedApiVersionCmd);
+        if (!supportedApiVersionAnswer.getResult()) {
+            throw new CloudRuntimeException("Incorrect API version: Nuage plugin only supports " + getClientDefaultsAnswer.getCurrentApiVersion());
+        }
+
+        String apiRelativePath = "/nuage/api/" + command.getApiVersion();
+        if (!apiRelativePath.equals(nuageVspHost.getDetails().get("apirelativepath"))) {
+            paramsTobeUpdated.put("apirelativepath", apiRelativePath);
+            paramsTobeUpdated.put("apiversion", apiVersion);
+        }
+
+        if (command.getApiRetryCount() != null &&
+                command.getApiRetryCount() != Integer.parseInt(nuageVspHost.getDetails().get("retrycount"))) {
+            paramsTobeUpdated.put("retrycount", String.valueOf(command.getApiRetryCount()));
+        }
+
+        if (command.getApiRetryInterval() != null &&
+                command.getApiRetryInterval() != Integer.parseInt(nuageVspHost.getDetails().get("retryinterval"))) {
+            paramsTobeUpdated.put("retryinterval", String.valueOf(command.getApiRetryInterval()));
+        }
+
+        if (paramsTobeUpdated.size() > 0) {
+            Map<String, String> latestParamsValue = new HashMap<String, String>();
+            latestParamsValue.putAll(nuageVspHost.getDetails());
+            latestParamsValue.putAll(paramsTobeUpdated);
+            Map<String, Object> hostdetails = new HashMap<String, Object>();
+            hostdetails.putAll(latestParamsValue);
+
+            try {
+                resource.configure("", hostdetails);
+                UpdateNuageVspDeviceCommand cmd = new UpdateNuageVspDeviceCommand(latestParamsValue);
+                UpdateNuageVspDeviceAnswer answer = (UpdateNuageVspDeviceAnswer)_agentMgr.easySend(nuageVspHost.getId(), cmd);
+                if (answer == null || !answer.getResult()) {
+                    s_logger.error("UpdateNuageVspDeviceCommand failed");
+                    if ((null != answer) && (null != answer.getDetails())) {
+                        throw new CloudRuntimeException(answer.getDetails());
+                    }
+                }
+                _hostDetailsDao.persist(nuageVspDevice.getHostId(), paramsTobeUpdated);
+            } catch (Exception e) {
+                throw new CloudRuntimeException("Failed to update NuageVsp device information", e);
+            }
+        } else {
+            s_logger.debug("No change in the NuageVsp device parameters. So, none of the NuageVsp device parameters are modified");
```

Comment:
Wrap in a if (s_logger.isDebugEnabled()) { ... } block to avoid unnecessary pressure on the string pool when debug logging is not enabled.